pkg/pillar/cmd/client/client.go

```diff
@@ -284,25 +296,68 @@ func Run(ps *pubsub.PubSub) { //nolint:gocyclo
 				// XXX keep exponential unchanged?
 				break
 			}
-			// XXX server/tls setup from above to capture change
-			// to server file?
-			if operations["selfRegister"] {
-				done = selfRegister(zedcloudCtx, onboardTLSConfig, deviceCertPem, retryCount)
+
+			// if device is allowed to be V2, get server certs first unless the cloud does not support V2 or the
+			var delay time.Duration
+			certRetryCnt := 0
+			for !gotServerCerts && !cloudV1Only && devMaybeV2 {
```

Comment:
You have to rewrite this to be a timer in an event loop the same way that getUuid and selfRegister are now done (when you started then might have been done with a sleep). That is required so that client will pick up new IP addresses while it is retrying; otherwise you could be retrying even though nim hasn't yet provided an IP address (which happens when the environment has a slow DHCP server which we see in some customer deployments),

Reply:
But I'm sharing the same timer used by getUuid and selfRegister. So it is waiting for the nim ip address update also, it's just this fetchCertChain runs first before the selfRegister and getUuid. Do I miss something here?

Reply:
oh, maybe i should not do the retry here internally? just noticed the getUuid does not do retry with a delay, will remove that.